### ISSUE TITLE
Better seeding for Random

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#65]
Currently it creates its randomness source using the default constructor of the `Random` class. But the well known problem with that is that this is seeded with the system time, which typically doesn't change for 16ms. So all instances of the random class created during that interval will return the same sequence.

My suggestion is creating one static instance of random and using it(with locking of course) to create the seed for the individual instances of `Random`.

---

Originally reported on Google Code with ID 65

Reported by `CodeInChaos@quantentunnel.de` on 2011-03-02 21:07:16

**Comment by fsateler** [comment]
I disagree with this request. [There is an overload that takes a `Random` parameter](https://morelinq.github.io/2.2/ref/api/html/M_MoreLinq_MoreEnumerable_Random_3.htm), so this should be fairly easy to implement in your own program. Moreover, not all programs need `Random` sequences to be different in quick sucession bursts. This would thus slow down `Random` for these users for no gain.

**Comment by fsateler** [comment]
I'm going to close this. Please reopen if you disagree with me.

**Comment by atifaziz** [comment]
@fsateler I think the issue is a valid concern because while [there's an overload that accepts a `Random`](https://morelinq.github.io/2.2/ref/api/html/M_MoreLinq_MoreEnumerable_Random_3.htm) as you point out, `Random` isn't thread-safe so it's not like you can put it in a static and supply the same instance from multiple threads to get _better seeding_. If we don't want to get bogged down with locking, the best balance here might be to go with a lazy-initialized thread-static so there'll be a `Random` per thread. This is non-trivial to ask everyone to get right and so we should just do it for the simpler overloads that instantiate and use a new `Random` on behalf of the user. A thread-static will also have the least impact on the existing implementation.

Another advantage of sharing a `Random` is that the initialization cost ([space](https://github.com/dotnet/corefx/blob/v1.0.4/src/System.Runtime.Extensions/src/System/Random.cs#L38) and [time](https://github.com/dotnet/corefx/blob/v1.0.4/src/System.Runtime.Extensions/src/System/Random.cs#L57-L86)) and while this may be lesser of a motivation, it's just another bonus.

Since there's only gain, I'd like to re-open and take a crack at this.

> This would thus slow down `Random` for these users for no gain.

You didn't say why this is slow but I'm guessing you were thinking locking?

**Comment by fsateler** [comment]
> I think the issue is a valid concern because while there's an overload that accepts a Random as you point out, Random isn't thread-safe so it's not like you can put it in a static and supply the same instance from multiple threads to get better seeding. 

I was thinking you would have a wrapper like this:

```csharp
static readonly Random  MyGlobalRandom = new Random();

public static IEnumerable<int> MyRandom() {
    Random newRand;
    lock(MyGlobalRandom) {
        newRand = new Random(MyGlobalRandom.Next());
    }
    return MoreEnumerable.Random(newRand);
}
```

> If we don't want to get bogged down with locking, the best balance here might be to go with a lazy-initialized thread-static so there'll be a Random per thread. This is non-trivial to ask everyone to get right and so we should just do it for the simpler overloads that instantiate and use a new Random on behalf of the user. A thread-static will also have the least impact on the existing implementation.

How would a thread-static variable interact with async contexts? I think it does not break (because there is no need to enforce the same `Random` seeder for each successive call), but I'm not very sure about how these contexts work.

However, a more problematic issue is that a thread-static `Random` seeder would have the same "problem" as currently in multi-threaded applications (like ASP.NET apps), as a request burst could initialize several threads with the same seed time.

> > This would thus slow down Random for these users for no gain.
>
>You didn't say why this is slow but I'm guessing you were thinking locking?

Yes. I was imagining an implementation like the above, which would incur the locking cost.

**Comment by atifaziz** [comment]
> as a request burst could initialize several threads with the same seed time.

You would have to seed each `Random` explicitly and not rely on the system time ticks anymore.

**Comment by fsateler** [comment]
>> as a request burst could initialize several threads with the same seed time.
>
>You would have to seed each Random explicitly and not rely on the system time ticks anymore.

With what? One could push the problem one step further back, and use a globally-locked `Random` used for creating the thread-level `Random` object, as described by this [blog post](https://blogs.msdn.microsoft.com/pfxteam/2009/02/19/getting-random-numbers-in-a-thread-safe-way/). That would reintroduce locking, but only for the first `Random` invocation on each thread.

**Comment by atifaziz** [comment]
@fsateler I've opened PR #265 with an initial draft. We can review, debate & improve the implementation there? And if we come to the conclusion that it's too risky to get it right then we can close it. Best to err on the safe side? I mean, as you said, there's an overload taking a `Random` so we could just leave it as an exercise for the user. 😉

**Comment by atifaziz** [comment]
> With what? 

I was initially thinking the [thread ID](https://msdn.microsoft.com/en-us/library/system.threading.thread.managedthreadid) but those recycle and they'd have to recycle ludicrously quickly. :smile:

![speed](https://cloud.githubusercontent.com/assets/20511/24256253/f37c909c-0fe7-11e7-8fe0-e36a23d5a265.jpg)

**Comment by atifaziz** [comment]
Closed by a36ba54d1dccd7165d4b5935ef4a54b48f73495d
